packages/workspace/src/browser/workspace-service.ts

```diff
@@ -16,48 +16,79 @@ import { WorkspaceServer } from "../common";
 @injectable()
 export class WorkspaceService {
 
-    /**
-     * The current workspace root.
-     */
-    readonly root: Promise<FileStat>;
-
     constructor(
         @inject(FileSystem) protected readonly fileSystem: FileSystem,
         @inject(FileSystemWatcher) protected readonly watcher: FileSystemWatcher,
         @inject(WorkspaceServer) protected readonly server: WorkspaceServer
-    ) {
-        this.root = this.server.getRoot().then(uri =>
-            this.validateRoot(uri)
-        );
-        this.root.then(root =>
-            watcher.watchFileChanges(new URI(root.uri))
-        );
+    ) { }
+
+    /**
+     * The promise which will resolve to the currently selected workspace root.
+     * This promise resolves to the workspace root file stat, if [rootResolved](WorkspaceService.rootResolved) is `true`.
+     */
+    get root(): Promise<FileStat> {
+        const self = this;
+        return new Promise((resolve, reject) => {
+            (function waitUntilResolved() {
+                self.server.getRoot().then(uri => {
+                    if (uri) {
+                        self.validateRoot(uri).then(
+                            root => resolve(root),
+                            () => setTimeout(waitUntilResolved, 200)
```

Comment:
Would it be possible to use JSON-RPC notification instead of polling?

Reply:
If you can provide an example, sure, I will change. Thanks!

Reply:
I mean to add `WorkspaceClient` similar to `ILoggerClient`: 
```ts
    constructor(
        @inject(FileSystem) protected readonly fileSystem: FileSystem,
        @inject(FileSystemWatcher) protected readonly watcher: FileSystemWatcher,
        @inject(WorkspaceServer) protected readonly server: WorkspaceServer
    ) {
        this.root = server.getRoot().then(current =>
            current ? this.validateRoot(current) : new Promise(resolve => {
                server.setClient({
                    onDidChangeRoot: uri => {
                        if (uri) {
                            resolve(this.validateRoot(uri));
                        }
                    }
                });
            )
        );
    }
```

Reply:
Thanks for the snippet! Well, yeah. Could be. Exactly this is what I wanted to avoid, boilerplate client, proxy, server, binding, blabla... just for listening to a single property change.

Reply:
Should we start using polling when we have 2 directional connection?

Reply:
> Would it be possible to

:) Sure, then in the future please write: _Replace it with JSON-RPC notification... Here comes an example snippet. `// snippet...` Let me know if it needs clarification._

And we can save some time and energy. Thanks!